actioncable/app/assets/javascripts/action_cable/connection.coffee

```diff
@@ -89,19 +97,49 @@ class ActionCable.Connection
 
     open: ->
       ActionCable.log("WebSocket onopen event")
-      @disconnected = false
-      @subscriptions.reload()
+      # This is in the case that the browser client opens
+      # the connection regardless of whether a protocol was 
+      # negotiated successfully or not.
+      if @isProtocolSupported()
+        ActionCable.log("WebSocket using '#{@getProtocol()}' subprotocol.")
+        @disconnected = false
+        @subscriptions.reload()
+      else
+        @close()
 
-    close: ->
+    close: (event) ->
       ActionCable.log("WebSocket onclose event")
-      @disconnect()
+
+      # According to https://www.w3.org/TR/websockets/#concept-websocket-close-fail
+      # a sub-protocol mismatch should be treated as a 1006 error, indistinguishable
+      # from any other failure to connect. Not all browsers behave this way:
+
+      # This encompases the behavior of most browsers when it comes to
+      # reacting to a failed protocol negotiation.
+      if event.code is 1000 and event.wasClean and @getProtocol() is ""
+        @disconnect(ActionCable.Events.Disconnected.ProtocolNotSupported)
+
+      # This is Chrome's response (which is technically to spec) and looks 
+      # exactly the same as a normal bad disconnect. Thus, we must test to 
+      # see if the server would connect under normal conditions. If so, Chrome reacted
+      # to the failed protocol negotiation, and not something else.
+      else if event.code is 1006 and not event.wasClean and @getProtocol() is ""
+        testSocket = new WebSocket(@consumer.url)
+        testSocket.onopen = =>
+          @disconnect(ActionCable.Events.Disconnected.ProtocolNotSupported)
+          testSocket.close()
```

Comment:
Well we can't send a `Sec-Websocket-Protocol` and then not receive one back from the server: the browser websocket clients just force disconnect. So that would mean moving the negotiation to outside of the websocket layer, which involves changing the protocol, and itself becomes subject to changes. :-)

Closing the connection is no problem: the browser is forcefully doing that for us. This whole thing is quite effective in preventing the client side from getting into a bad state, which is the primary concern.

But the issue that remains is we can't tell the app why there has been this specific disconnect in Chrome, unless we use this reconnect thing. For me, that's not ideal but it's not a deal breaker.

I see a few options:

1.  Keep the double reconnection thing. It's going to result in Chrome doing two very short-lived connections, but only if the protocol changes. However, the connection attempts will stop after that.
   **Pros**: We get consistent behavior across all browsers; from a load perspective it will in aggregate be less (no retries).

  **Cons**: _Only_ when there is a protocol change, there will be two back to back connection attempts by Chrome users; Unclear what load profile this would create. Solution is non standard.
2. Let Chrome continue to retry as it normally would.
  **Pros**: No back to back connection attempt for Chrome.

  **Cons**: Inconsistent behavior across browsers, and Chrome users won't be able to get a user-friendly message to refresh. Chrome will continue to retry fruitlessly.
3. Don't attempt to understand reason for connection and continue retrying, but the app won't know why it disconnected.
  **Pros**: Consistent across all browsers.

  **Cons**: No nice message to users. All browsers are going to continue to retry fruitlessly.

4. Move protocol negotiation out of websocket layer.

  **Pros**: We can get consistent behavior across browsers.
 
 **Cons**: Re-inventing the wheel; solution is not resistant against future protocol changes; binds ActionCable to a content encoding.

Personally I think 1 or 3 are the best.

@jeremy Any opinions on this?